Improve CoqIDE documentation

jfehrle opened the pull request:
Filling in gaps in the presentation.

@Alizter Would you care to review this?  More generally, would you be interested in being a documentation reviewer? That would include reviewing for writing, formatting, organization, correctness and content--and getting help on correctness and content from subject matter experts as needed (when possible).

BTW, ".. todo PR" indicates things that should be fixed during the review process.

jfehrle commented:
Usually it takes several incremental reviews and further revisions before merging.  For a first review, think about organization and what's missing.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+The upper right panel is the *goal panel*, which shows
+the goals to be proven.  The lower right panel has the *message
+panel*, which shows messages produced by commands and tactics.
+The contents of the right-hand panels are specific to the currently-displayed script.
```
There are also tabs for different outputs. Perhaps these should be detailed.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+the goals to be proven.  The lower right panel has the *message
+panel*, which shows messages produced by commands and tactics.
+The contents of the right-hand panels are specific to the currently-displayed script.
+The status bar is at the bottom.
```
```suggestion
The *status bar* is displayed at the bottom.
```

jfehrle replied:
"displayed" isn't really needed, right?

Alizter replied:
As you wish.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+-----------------------------------------
+
+The *File* menu lets you open files, create new buffers, save buffers to a file,
+and print or export them into various formats.
```
```suggestion
and print or export them in various formats.
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+The *File* menu lets you open files, create new buffers, save buffers to a file,
+and print or export them into various formats.
+
+Text editing provides the basic operations such as copy/cut/paste and replace.
```
```suggestion
Text editing provides the basic operations such as copy, cut, paste, find and replace.
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+
+Text editing provides the basic operations such as copy/cut/paste and replace.
+Most editing operations are shown in the *Edit* menu, which shows the equivalent
+keystrokes for operations.  if you need more complex editor commands, you can
```
```suggestion
keystrokes for operations.  If you need more complex editor commands, you can
```

Perhaps a note about the keystrokes appearing on the right of the operation in a faded font. (Watermark?)

jfehrle replied:
Got it, but above.  I think watermark is not the right word for that.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+Most editing operations are shown in the *Edit* menu, which shows the equivalent
+keystrokes for operations.  if you need more complex editor commands, you can
+launch your favorite text editor on the current buffer, using the
+*Edit/External Editor* menu. (When you're done editing you must currently must
+reopen the file to see your changes.)  Also note:
```
```suggestion
*Edit/External Editor* menu. When you're done editing you must currently must
reopen the file to see your changes.  Also note:
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+*Edit/External Editor* menu. (When you're done editing you must currently must
+reopen the file to see your changes.)  Also note:
+
+- Undo is also available as Ctrl-Z.  Redo is Ctrl-Shift-Z.
+- Select all is Ctrl-A.
+- Commenting and uncommenting the current line or selected text is available in
+  the *Tools* menu.
```
Ctrl (Linux and Windows) and Cmd for Mac I believe. Perhaps we should simply have a note somewhere about Mac users.

Alizter replied:
Also commenting and commenting is mapped to Ctrl-D and depends on the line of the cursor and the text highlighted.

jfehrle replied:
We should find a Mac user to verify this.  I didn't mention Ctrl-D because the mapping is discoverable from the menu (while Ctrl-Z and Ctrl-A don't appear on any menu).

jfehrle replied:
I think Ctrl-A and Ctrl-Z may be hard coded into the script panel widget.

Per https://support.apple.com/en-us/HT201236, these should be equivalent to Command-A and Command-Z on macos.

FWIW, we need a documented user-friendly way to configure key mappings.  I couldn't get https://github.com/coq/coq/wiki/Configuration-of-CoqIDE to work so far.  You've seen the Edit/Preferences/Shortcuts panel (which works)?  But not in this PR.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+`foo.v` is saved to `#foo.v#` every 10 seconds by default.  You can change the
+interval in the *Edit / Preferences / Files* dialog.
+
+In the *View* menu, you can set almost a dozen printing options that
```
11 to be exact. However I feel as though dozen might be more familiar to native English speakers. Perhaps "several" would work better.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+interval in the *Edit / Preferences / Files* dialog.
+
+In the *View* menu, you can set almost a dozen printing options that
+correspond to options that can appear in the script.  For example, "Display
+notations" on the menu corresponds to the :flag:`Printing Notations` flag.  You
+should use the menu instead of controlling these settings in your script.
```
Do we want to link to the Options/Flags section of the refman?

jfehrle replied:
``` :flag:`Printing Notations` ``` hyperlinks to that specific flag.  I don't think we can link to the flags options index, and the general explanation of flags and options is too basic.  Unfortunately, the menus don't exactly match the flag names, though they are close.

Alizter replied:
Some of the options set here, set multiple options IIRC. Honestly, I found this entire feature poorly executed, since it also explicitly stops you from setting options within the buffer.

Alizter replied:
Reported as https://github.com/coq/coq/issues/16080

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+Operations for running the script are available in the *Navigation* menu,
+from the toolbar and from the keyboard.  These include:
+
+- Forward (Ctrl-Down) to run one command or tactic
+- Backward (Ctrl-Up) undo one command or tactic
+- Run to cursor (Ctrl-Right) to run commands up to the cursor
+- Run to end (Ctrl-End) to run commands to the end of the buffer
+- Reset Coq (Ctrl-Home) to restart the Coq process
+- Interrupt to stop processing commands
```
End these in periods or no?

jfehrle replied:
No.  They are sentence fragments.  (Yes if they were sentences.)

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+- Reset Coq (Ctrl-Home) to restart the Coq process
+- Interrupt to stop processing commands
+
+Commands may:
```
Some screenshots may help here?

jfehrle replied:
I added that the toolbar icons have tooltips.  The users can figure it out.  (An exercise for the reader.)  Screenshots take a while to get right, are somewhat system-dependent and quickly get out of date, so I use them sparingly.  For example, compare the first 2 screenshots in the chapter with the third--fairly different.  The first two are old, the third is quite new.

jfehrle replied:
The triage principle applies equally to the documentation--address the most important things first.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+  Double click on an entry to jump to the point of the error.  Execution
+  of commands stops unless you're in async mode.
+
+.. todo: describe the ribbon (a bit broken, though)
```
How broken? I think the ribbon is a very useful indicator on the progress of a proof and the fact that you can click it to navigate the buffer is quite useful.

jfehrle replied:
I don't know how to explain how to interpret it and its dynamic behavior.  It's animated but doesn't communicate very well.  For example, looking at Pff.v, first I see

![image](https://user-images.githubusercontent.com/1253341/170413436-dcfb689a-b1b3-4b11-b4eb-bcb16f744b67.png)

then a bit later I see

![image](https://user-images.githubusercontent.com/1253341/170413500-02f0fbb5-ac6c-4908-bd54-c60ca21daf36.png)

What do the gray lines mean, eg in Datatypes.v:

![image](https://user-images.githubusercontent.com/1253341/170414203-34cec18b-b664-46c1-b81c-7ffd73d33ebf.png)

It seems to me it's a somewhat unfinished feature.

Alizter replied:
I would think that `Pff.v` is quite an unusual file due to being over 1000 lines, and I would agree that the ribbon feature in this case is not very helpful. It can be confusing that the ribbon only matches what Coq has actually read, and that the rest of the file is still being read. I don' know if there is an easy way to communicate that.

The grey dots are the unchecked Qeds, and should go away after pressing the check button. I believe this is also not the correct colour. This is probably because the colours in the ribbon are hard coded rather than being derived from the style.

For a 300 line file, for instance. It can be a useful indicator, especially if you have any `Axiom`s or `Admitted` proofs since they show up in yellow.

Alizter replied:
@jfehrle Would you like to write a little bit about the ribbon however? I think mentioning that it is clickable might be useful.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
-If you ever try to execute a command that runs for a long time
-and would like to abort it before it terminates, you may
-use the interrupt button (the white cross on a red circle).
+If you're not in async mode and you modify the processed part of the buffer,
```
Link to async?

Alizter replied:
Or maybe not it comes quickly after. Perhaps instead we can say, see further down, or something.

jfehrle replied:
I did that on about 10 lines above this.

BTW, do you know how to access the html doc for a specific PR?

Alizter replied:
Yes, I was reviewing on my phone so I could only read the diff. I will do another pass in HTML.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
-and would like to abort it before it terminates, you may
-use the interrupt button (the white cross on a red circle).
+If you're not in async mode and you modify the processed part of the buffer,
+everything after that point is undone.  Unlike `coqtop`, you should not use
+:cmd:`Undo` to go backward.
```
Do we ever recommend the use of `Undo` and/or `coqtop` in the refman?

jfehrle replied:
No, IIRC it doesn't work correctly at all.  A couple examples use it, perhaps they should be changed.

Alizter replied:
I recall vaguely that there was a plan to deprecate and eventually remove it completely. I don't know what happened to that. I would think that the other devs are not particularly fond of keeping it around.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+Async mode
+----------
+
+Asynchronous mode uses multiple Coq processes to process elements of the script
+in parallel.  For example, multiple theorems may be processed in parallel.  This is
```
I would say "concurrent" rather than "parallel". Concurrent would mean that the application is processing many separate tasks at the same time, whilst parallel would indicate there is one task broken down into smaller subtasks.

Alizter replied:
Then again maybe we made a mistake with calling it "parallel proof processing". I suppose the task depends on your point of view.

jfehrle replied:
It is parallel (when things go well :-) ).  Coq creates multiple proof worker processes that can run on different cores.  Not merely concurrent.

Alizter replied:
Right, but my point was that they are not processing the same proof, but separate proofs in the same buffer. Minor point anyway.

jfehrle replied:
The "multiple theorems" can be processed in parallel, but individual theorems can't be split and parallelized as things are now. (Maybe different in the future).  I think that would be the most obvious understanding from the current wording.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+in parallel.  For example, multiple theorems may be processed in parallel.  This is
+described in detail in :ref:`asynchronousandparallelproofprocessing`.
+
+While synchronous mode stops processing at the first error it encounters, in async
+mode, errors only stop processing the proof the error appears in.
```
Do we want to mention that a "proof" is delaminated by `Proof.` and `Qed.`? I am not even sure `Proof.` is even needed.

jfehrle replied:
Got it. I think you mean "delimited", eh?  I was told `Proof.` was required.  If I say it's required and it's not, I doubt anyone will be confused or will complain.

Alizter replied:
`Proof` is recommended but not required. This will likely change in the future, but there is still some pushback from users that do not want to write 6 characters.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
 
-Notice that for all these buttons, except for the "gears" button, their operations
-are also available in the menu, where their keyboard shortcuts are given.
+In the screenshot, the proof of the failed theorem can be edited (between `Proof.`
+and `Qed.`) without invalidating the theorems that follow it.  Then the modified
+proof can be reprocessed using the usual navigation operations.  The light blue
+highlight in the script indicates commands that haven't been processed.
```
```suggestion
In the screenshot, the proof of the failed theorem can be edited (between `Proof.`
and `Qed.`) without invalidating the theorems that follow it.  The modified
proof can then be reprocessed using the usual navigation operations.  The light blue
highlight in the script indicates commands that haven't been processed.
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+proof can be reprocessed using the usual navigation operations.  The light blue
+highlight in the script indicates commands that haven't been processed.
+
+Async mode defers the final type checking step of proofs, leaving the `Qed.` marked in
+a slightly different shade of light blue to indicate this.  To complete the final
+checking, click on the "gears" button on the toolbar ("Fully check the document").
```
Wow, I only just now found out about this.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+:cmd:`Check`, :cmd:`Search`, etc.  The *query pane* lets you run such commands
+interactively without modifying your script. The query pane is accesible from
+the *View* menu, or using the shortcut ``F1``.
+You can also do queries by selecting a particular phrase, then choosing an
+item from the *Queries* menu. The response will appear in the message panel.
```
```suggestion
You can also do queries by selecting a particular phrase, then choosing an
item from the *Queries* menu. The response will appear in the message panel.
```

Selecting a phrase or highlighting text with the cursor? Phrase doesn't have any technical meaning IIRC.

Alizter commented:
@jfehrle I would be interested in joining the documentation team.

jfehrle commented:
> @Alizter I would be interested in joining the documentation team.

@Zimmi48: Would you add @Alizter?

jfehrle commented on doc/sphinx/addendum/parallel-proof-processing.rst:
```diff
@@ -112,6 +112,9 @@ Valid proof block types are: “curly”, “par”, “indent”, and “bullet
 Interactive mode
 ---------------------
 
+.. todo: still only CoqIDE?  Trim this back a bit, some of the info is now
+   in the CoqIDE chapter.
```
Who can confirm that?  (Even so, we don't want to say too much about the other IDEs.)

Alizter replied:
https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/CI.20for.20base.2Basync.20timing.20out/near/283809250

https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/CI.20for.20base.2Basync.20timing.20out/near/283861384

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+
+The Coq Integrated Development Environment is a user-friendly GUI
+replacement for `coqtop`. Its main purpose is to
+allow users to edit scripts and step forward and backward through Coq scripts,
```
British English often uses "forwards", American English not so much.  For better or worse, we've chosen to use American English for the documentation (fine with me, of course, but I was not involved in the choice).

See https://www.collinsdictionary.com/dictionary/english/forward

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+allow users to edit scripts and step forward and backward through Coq scripts,
+respectively executing and undoing commands and tactics.
+
+To run CoqIDE, enter `coqide` in operating systems's command line.
```
Best not to get into environment-specific details like starting coqide through shortcuts.  I expect the users who care know how to create them.

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
+
+To run CoqIDE, enter `coqide` in operating systems's command line.
+If you include the script file names as arguments, each is opened
+in a separate tab.  If you don't, CoqIDE opens a single unnamed buffer
+(`*scratch*`).  `coqide` also accepts many of the options of `coqtop`
```
"**unnamed** buffer **named** (`*scratch*`)" doesn't read very well...

Alizter replied:
How about "titled"?

jfehrle replied:
Wait until I push an update and a comment to that effect.  Feedback while I'm in the midst of updating gets really confusing.

jfehrle replied:
OK, titled.  I had used "labelled".

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,165 @@
-goals to be proven are displayed. The lower right window is the *message
-window*, where various messages resulting from commands are displayed.
-At the bottom is the status bar.
+The screen shot shows CoqIDE while the user is stepping through the file `Fermat.v`.
```
"Whilst" is an Anglicism.  See https://www.grammarly.com/blog/while-vs-whilst/

jfehrle commented:
Updated.  I pushed a new commit addressing your comments.

When I feel I've fully applied one of your comments, I resolve it.  I leave it unresolved if I added my comments.  Feel free to resolve those if you're satisfied.

Still some more changes to make...

Alizter commented:
@jfehrle Some error in the refman build, I will review the HTML when you have fixed it.

jfehrle commented:
> Some error in the refman build, I will review the HTML when you have fixed it.

I forgot to check in screen shot image.  It should build now.

jfehrle commented:
Updated.  Some new material in the newest commit.  I don't plan to add any more material.  I will squash before you merge.

Alizter commented:
@jfehrle Something broke with your last push.

jfehrle commented:
Got it, updated.  Sphinx is a bit error prone.

Alizter commented:
I wonder if we can insert "CoqIDE" into the title here: https://coq.gitlab.io/-/coq/-/jobs/2518205745/artifacts/_install_ci/share/doc/coq/sphinx/html/practical-tools/coqide.html?highlight=coqide#coq-integrated-development-environment

If I search CoqIDE in the search bar of the refman for example, this isn't the first hit. If the title said:
```
CoqIDE - Coq Integrated Development Environment  
```
or even just
```
CoqIDE
```
That would be better in my opinion. The explanation of IDE can come in the main body of the text if needed. As a title it seems redundant.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+.. todo PR: how to say that a number of things are broken?  Maybe list them
+   somewhere--doesn't have to be super detailed
+
+The Coq Integrated Development Environment (CoqIDE) is a user-friendly GUI
+for Coq. Its main purpose is to
+allow users to edit scripts and step forward and backward through Coq scripts,
+respectively executing and undoing commands and tactics.
```
This paragraph is hard to read IMO.  
```suggestion
The Coq Integrated Development Environment (CoqIDE) is a user-friendly GUI
for Coq. Its main purpose is to
allow users to edit scripts and step through Coq scripts, executing and undoing commands and tactics.
```

The respectively part encourages me to reread, but who knows, maybe I can't read.

Alizter replied:
Also please rehard wrap.

jfehrle replied:
Rewrapping is a manual process.  My GUI doesn't have a command to do that.  Also it doesn't matter, does it?  We expect users to read the HTML.  But I did rewrap this couple of lines.

Alizter replied:
@jfehrle My GUI allows me to do some rewrapping, it can help when reading the rst files directly, which is sometimes done. It also helps people when they edit it later. Not a big deal if you cannot do it right now however.

jfehrle replied:
It's OK to rewrap the changed parts, but en masse rewrapping would obscure the git history.  I expect this is not news to you.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+respectively executing and undoing commands and tactics.
+
+To run CoqIDE, enter `coqide` on the command line.
+If you include script file names as arguments, each is opened
```
Do we mention that "script file" = ".v file"? We should perhaps include (.v) or something.

jfehrle replied:
I'm sure it's in the document somewhere but some repetition is fine.

jfehrle commented:
I'll make it "CoqIDE".

BTW, do you know anything about the lint failures (all related to Stream), e.g. if someone is working on it and/or whose changes may have caused it?  Not a blocker for this PR, but if it's very easy to fix I might give it a crack.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
-goals to be proven are displayed. The lower right window is the *message
-window*, where various messages resulting from commands are displayed.
-At the bottom is the status bar.
+The screen shot shows CoqIDE while the user is stepping through the file `Fermat.v`.
```
```suggestion
The screen shot shows the interface of CoqIDE as the user is stepping through the file `Fermat.v`.
```

jfehrle replied:
Does "the interface of" add really anything?  I think it just makes the text wordier.  I'm a big fan of being succinct (subject to also being clear).

Alizter replied:
OK then don't worry about the first part.

Alizter commented:
@jfehrle Yep, that's because we bumped to OCaml 4.14 in the CI which alerts when the Stream library is used. If you rebase, it should go away.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
-window*, where various messages resulting from commands are displayed.
-At the bottom is the status bar.
+The screen shot shows CoqIDE while the user is stepping through the file `Fermat.v`.
+A menu bar and tool bar appear at the top of the window. The left-hand panel shows
```
```suggestion
A menu bar and a tool bar appear at the top of the window. The left-hand panel shows
```
I'm not sure about this one however. But I'll state it in case you agree.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+panel*, which shows messages produced by commands and tactics; the *Errors panel*,
+which shows errors detected when running in :ref:`async mode <asyncmode>` and the *Jobs panel,* which shows
+information on the worker processes used by async mode.
+The contents of the right-hand panels are specific to the currently-displayed script.
```
We should also mention the "pop-out" feature, which lets you pop the tabs out in their seperate window, to be restored when the window is closed.

jfehrle replied:
We call it "detached" (as in `Windows/Detach proof` on the menu and the `Detach` buttons on the Query panel and debugger panels.

Alizter replied:
Yep, detached is the correct word. Will you mention it here then?

jfehrle replied:
Yup.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+Text editing provides the basic operations such as copy, cut, paste, find and replace.
+Most editing operations are shown in the *Edit* menu.  Most menu entries show the equivalent
+keystrokes to perform the action.  If you need more complex editor commands, you can
+launch your favorite text editor on the current buffer, using the
```
Your "favorite text editor" is configurable in the preferences menu.

jfehrle replied:
While `Edit/External editor` saves the buffer if it's modified, IMHO when that editor exits, CoqIDE should reload the file if has been changed.  We also should have a way for CoqIDE to detect and act on external changes to .v files, .vo files, etc.

The external editor fails if you're running in from the source tree.  Also there should be a message if the editor doesn't come up.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+*Edit/External Editor* menu. When you're done editing, you must currently must
+reopen the file to see your changes.  Also note:
+
+- Undo is also available as Ctrl-Z (Command-Z on macOS).  Redo is Ctrl-Shift-Z
+  (Shift-Command-Z on macOS).
```
MacOS shortcuts are typically abbreviated to Cmd rather than Command. The same way Control is abbreviated to Ctrl on Windows and Linux. Mac keyboards (that I have seen) have Cmd written on them.

jfehrle replied:
The Apple web page I consulted had it spelled out.  But "Cmd" is an abbreviation similar to "Ctrl" and "Alt", so more appropriate.

Alizter replied:
Here is how Apple shortens it:

![image](https://user-images.githubusercontent.com/8614547/170846231-9faf5881-666a-4add-b947-3b978dece1e5.png)

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+  (Shift-Command-Z on macOS).
+- Select all is Ctrl-A (Command-A on macOS).
+- Ctrl-Delete deletes a word of text after the cursor
+- Ctrl-Backspace deletes a word of text before the cursor
```
Perhaps for these keyboard shortcuts, we should use code blocks to better distinguish from the text.

jfehrle replied:
Done.  I think we need a font that's more distinctive (later on):

![image](https://user-images.githubusercontent.com/1253341/170846029-9534ba85-74a0-49f3-acd0-d3657bff66ef.png)

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+- Run to cursor (Ctrl-Right) to run commands up to the cursor
+- Run to end (Ctrl-End) to run commands to the end of the buffer
+- Reset Coq (Ctrl-Home) to restart the Coq process
+- Interrupt to stop processing commands
```
In my experience, if CoqIDE is very busy, for example printing a huge term, interrupt can be somewhat unresponsive. The goal is to not undo the processed part of the script and just interrupt the current command. If CoqIDE is really not responsive however, resetting Coq with Ctrl-Home typically does the job.

I don't know if this needs mentioning.

jfehrle replied:
Interrupt is implemented with a signal sent directly to coqidetop that sets a flag.  The interpreter checks the flag before starting the next command.  I tweaked the wording.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+
+Commands may:
+
+- Complete successfully.  In this case, the background of the command is marked
+  with the "processed" color (green by default).
+- Complete with a warning.  In this case, the warning appears in the messages
+  panel in yellow.  The background of the command is marked with the "processed"
+  color and the text is shown in blue and underlined.  The message text is
+  available as a tooltip on the text of the command.
+- Fail with an error.  If you're stepping through the proof line by line, the
+  error message appears in the message panel in red and the command is shown
+  in red and underlined with a pink background.  If you're in async mode,
+  described in more detail below, the message appears in the *errors panel*.
+  Double click on an entry to jump to the point of the error.  Execution
+  of commands stops unless you're in async mode.
```
Also here, we have Axiom's and Admitted which are highlighted in yellow in the buffer.

jfehrle replied:
Think I'd call it light orange.  Different shades in the script and in the ribbon BTW.  Done. but OTOH this is something that users will effortlessly observe (I don't recall ever being told about this behavior).

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+
+In the previous figure :ref:`CoqIDE main screen <coqide_mainscreen>`,
 the running buffer is `Fermat.v`, all commands until
 the ``Theorem`` have been already executed, and the user tried to go
```
```suggestion
the ``Theorem`` have already been executed, and the user has tried to go
```

jfehrle replied:
How about "all commands until the ``Theorem`` have already been executed, then the user tried to go"?

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
+In the previous figure :ref:`CoqIDE main screen <coqide_mainscreen>`,
 the running buffer is `Fermat.v`, all commands until
 the ``Theorem`` have been already executed, and the user tried to go
 forward executing ``Induction n``. That command failed because no such
```
```suggestion
forward executing ``Induction n``. That command has failed, since no such
```

jfehrle replied:
I think the simple past is fine.  You don't like "because" here?

Alizter replied:
OK fair enough, I think yours is simpler.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
 the ``Theorem`` have been already executed, and the user tried to go
 forward executing ``Induction n``. That command failed because no such
 tactic exists (names of standard tactics are written in lowercase),
 and the failing command is underlined.
```
```suggestion
and the failing command has been underlined.
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
 syntax of the command you want.
 
 Moreover, from this menu you can automatically insert templates of complex
 commands like ``Fixpoint`` that you can conveniently fill afterwards.
```
```suggestion
commands like ``Fixpoint`` that you can conveniently fill in afterwards.
```

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
-``Nat.mul`` in the script window, and choosing ``Print`` from the ``Queries``
-menu.
+A *query* is any command that does not change the current state, such as
+:cmd:`Check`, :cmd:`Search`, etc.  The *query pane* lets you run such commands
```
```suggestion
:cmd:`About`, :cmd:`Check`, `cmd:`Print`, :cmd:`Search`, etc.  The *query pane* lets you run such commands
```

jfehrle replied:
Got it.  Do we still need the "etc."?

Alizter replied:
Technically there are a few more so I would keep it.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -5,130 +5,183 @@
-----------------
+-----------
 
 The `Compile` menu offers direct commands to:
```
Shall we mention that this doesn't work at the moment, might get fixed in the future and that compiling from the terminal is a more sound alternative?

jfehrle replied:
In this case, yes.  It's on my list to fix. I expect it will be straightforward and that I'll get to it fairly soon.

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -138,14 +191,16 @@ The `Compile` menu offers direct commands to:
+Preferences
+~~~~~~~~~~~
+
+You may customize your environment with the *Preferences* dialog, which
+*Edit/Preferences* on the menu. There are several sections:
```
Which Edit/Preferences? Do you mean "which is found under Edit/Preferences"?

Alizter commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -184,20 +239,61 @@ Notice that these settings are saved in the file ``coqiderc`` in the
+
+.. todo: list common rebindings?
+
+.. todo: microPG mode?
```
I don't particularly remember, but I seem to be under the impression this enables the PG keybindings rather than the default CoqIDE ones.

jfehrle replied:
Yeah, I've not found time to experiment with it.  See `Help/Help for uPG mode` and `Preferences/Editor`.  On the former, I wonder what key `Meta` corresponds to.  Also what happens to the accelerator settings it clashes with.

jfehrle commented:
Updated.

jfehrle commented:
@Alizter Will you be able to finish this one soon?  Maybe just review the newest commit.  In your copious free time :-).

Alizter commented:
I've posted some more comments, but overall it is looking very good! When you are done this round, I can push a commit fixing the wrapping in the entire CoqIDE document, which you can then merge with your commits before pushing.

jfehrle commented on doc/changelog/09-coqide/16070-coqide_async_doc.rst:
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  Documentation of editing failed async mode proofs,
+  how to configure key bindings and some heretofore
+  undocumented details
```
Mmm, I think "of" is better.  And "various previously" (much as I like "heretofore" :-).

jfehrle commented on doc/sphinx/addendum/parallel-proof-processing.rst:
```diff
@@ -43,34 +43,34 @@ When a section ends, Coq looks at the proof object to decide which
 --------------------------------------
 
-Coq 8.6 introduced a mechanism for error resilience: in interactive
+In interactive
 mode Coq is able to completely check a document containing errors
 instead of bailing out at the first failure.
```
Error resilience is mentioned in the async proofs section.  IIUC Enrico's new Stm will make possible new behavior in the GUI, so not worth capturing every detail.  But I was hardly aware of the async behavior until just a couple months ago--I often run the native Windows version where it's turned off by default, also I'm not creating my own proofs.  I'm not a typical user.

> For example, if I step down to an error, I can progress no further unless I do a step down to cursor or end of file.

Clearly users will discover that on their own, so not important to cover in the doc.

jfehrle commented on doc/sphinx/addendum/parallel-proof-processing.rst:
```diff
@@ -112,19 +112,21 @@ Valid proof block types are: “curly”, “par”, “indent”, and “bullet
+.. todo: How about PG and coqtail?
 
-When CoqIDE is started, two Coq processes are created. The master one
+CoqIDE and vsCoq support asynchronous proof processing.
```
sb "VsCoq".  See the README in https://github.com/coq-community/vscoq.  Right?

Alizter replied:
It is capitalised differently in the VSCode store: https://marketplace.visualstudio.com/items?itemName=maximedenes.vscoq

jfehrle replied:
Only in the title.  "VsCoq" is used in 2 places in the text of that page.  "VS Code" (with a space) is used in other pages at that site though.  FWIW.

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -2,133 +2,208 @@
+which shows errors detected when running in :ref:`async mode <asyncmode>` and the
+*Jobs panel,* which shows information on the worker processes used by async mode.
+The contents of the right-hand panels are specific to the currently-displayed script.
+Click the arrow icons to detach these panel into separate windows.  The proof panel
```
I think the command form reads a lot better and more easily understood than the passive voice.  And more succinct.

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -2,133 +2,208 @@
+and print or export them in various formats.
+
+Text editing provides the basic operations such as copy, cut, paste, find and replace.
+Most editing operations are shown in the *Edit* menu.  Most menu entries show the equivalent
```
"Keystroke equivalents (if defined) for menu items are shown on the right of each item."

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -2,133 +2,208 @@
+Text editing provides the basic operations such as copy, cut, paste, find and replace.
+Most editing operations are shown in the *Edit* menu.  Most menu entries show the equivalent
+keystrokes to perform the action.  If you need more complex editor commands, you can
+launch your favorite text editor on the current buffer, using the
```
It was in the original text.  It was cute.  It is gone.

Alizter replied:
I thought it was cute too, though we cannot assume our readers will enjoy comparing text editors. 0:-)

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -2,133 +2,208 @@
+- Select all is `Ctrl-A` (`Cmd-A` on macOS).
+- `Ctrl-Delete` deletes a word of text after the cursor
+- `Ctrl-Backspace` deletes a word of text before the cursor
+- Commenting and uncommenting the current line or selected text is available in
+  the *Tools* menu.  If some text is selected, exactly that text is commented out;
+  otherwise the line containing the cursor is commented out.  To uncomment, position
+  the cursor between `(*` and `*)` or select any text between them.
```
I'm reluctant: a) it is on the menu and b) the user can reconfigure it.  If I just say "Ctrl-D" it could lead to puzzlement.

jfehrle commented on doc/sphinx/practical-tools/coqide.rst:
```diff
@@ -2,133 +2,208 @@
+  otherwise the line containing the cursor is commented out.  To uncomment, position
+  the cursor between `(*` and `*)` or select any text between them.
+
+Files are automatically saved periodically to a recovery file.  For example,
+`foo.v` is saved to `#foo.v#` every 10 seconds by default.  You can change the
+interval in the *Edit / Preferences / Files* dialog.
```
I'd like to get rid of the recovery files to enable a smoother workflow with compile and external make.  Instead, if possible, save the files automatically when CoqIDE loses focus--though I'm not sure I can get that to work with XWindows, at least on my system.  And when CoqIDE regains focus, it should reload any files that have changed.  Users should be using a source code control system, which I recommend in #15888.

FWIW my IDE, IntelliJ IDEA, has some fabulous undo and diffing capabilities.  For example, it takes automatic snapshots of every file as you edit it.  This info is then available as "Local History".  I can edit in the right hand panel.  Clicking on ">>" undoes a change.  (I have no intention of adding anything like this to CoqIDE.)

![image](https://user-images.githubusercontent.com/1253341/172024725-6dc2d4e4-4e48-40b2-aa16-09659c5b637c.png)

jfehrle commented:
Updated and squashed.

> When you are done this round, I can push a commit fixing the wrapping in the entire CoqIDE document, which you can then merge with your commits before pushing.

I don't want to reformat the entire file or files.  That erases too much git history and makes it impossible to pick out the real changes between commits before and after.  Also some lines can't be broken (e.g. `..tacn::` definitions), so it will require reviewing the reformatted text.  It's OK but not essential to reformat lines and sometimes paragraphs that have been substantially rewritten.  Ragged line breaks don't impair readability IMHO.

Alizter commented:
If you are happy with this I will merge tomorrow.

jfehrle commented:
Updated, I'm happy, please merge.  Thanks for the review and your suggestions!

Alizter commented:
@coqbot merge now

cc @ppedrot

Alizter commented:
@coqbot merge now pls
